File: Data Structures & Algorithms/longest-repeating-substring-with-replacement/submission-3.java
class Solution {
    public int characterReplacement(String s, int k) {
        if(s.length() < 1) {
            return 0;
        }
        int start = 0;
        int end = 1;
        int totalMax = 1;
        while(end < s.length()) {
            int[] freq = new int[26];
            int maxFreq = 0;
            for(char c: s.substring(start, end+1).toCharArray()) {
                int idx = (c - 'A');
                freq[idx] += 1;
            }
            
            for(int c: freq) {
                maxFreq = Math.max(c, maxFreq);
            }

            if(k < ((end - start + 1) - maxFreq)) {
                start++;
            } else {
                totalMax = Math.max(totalMax, (end - start + 1));
                end++;
            }

        }

        return totalMax;
    }
}

File: Data Structures & Algorithms/valid-parenthesis-string/submission-3.py
class Solution:
    def checkValidString(self, s: str) -> bool:
        maxLeft = 0
        minLeft = 0

        for c in s:
           
            if c == '(':
                maxLeft += 1
                minLeft += 1
            elif c == '*':
                minLeft -= 1
                maxLeft += 1
            if c == ')':
                maxLeft -= 1
                minLeft -= 1

            if maxLeft < 0:
                return False
            minLeft = max(minLeft, 0)
        
        if minLeft <= 0 and maxLeft >= 0:
            return True
        return False